# World Project Database

## imports for Python, Pandas

In [1]:
import pandas as pd
import numpy as np

In [2]:
import json
from pandas.io.json import json_normalize

****
## JSON exercise

Using data in file 'data/world_bank_projects.json' and the techniques demonstrated above,
1. Find the 10 countries with most projects
2. Find the top 10 major project themes (using column 'mjtheme_namecode')
3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

## Load JSON Data

In [3]:
# load json as string

data = json.load((open('data/world_bank_projects.json')))

In [4]:
# load as Pandas dataframe

json_df = pd.read_json('data/world_bank_projects.json')
json_df.head()

_id  approvalfy board_approval_month  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999             November   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015             November   
2  {'$oid': '52b213b38594d8a2be17c782'}        2014             November   
3  {'$oid': '52b213b38594d8a2be17c783'}        2014              October   
4  {'$oid': '52b213b38594d8a2be17c784'}        2014              October   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   
2  2013-11-01T00:00:00Z   MINISTRY OF FINANCE AND ECONOMIC DEVEL   
3  2013-10-31T00:00:00Z   MIN. OF PLANNING AND INT'L COOPERATION   
4  2013-10-31T00:00:00Z                      MINISTRY OF FINANCE   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   
2                   NaN                                   Tuvalu!$!TV   
3                   NaN                        Republic of Yemen!$!RY   
4  2019-04-30T00:00:00Z                       Kingdom of Lesotho!$!LS   

  countrycode                              countryname    countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia            Ethiopia   
1          TN                      Republic of Tunisia             Tunisia   
2          TV                                   Tuvalu              Tuvalu   
3          RY                        Republic of Yemen  Yemen, Republic of   
4          LS                       Kingdom of Lesotho             Lesotho   

   ...   sectorcode source  status  supplementprojectflg  \
0  ...  ET,BS,ES,EP   IBRD  Active                     N   
1  ...        BZ,BS   IBRD  Active                     N   
2  ...           TI   IBRD  Active                     Y   
3  ...           JB   IBRD  Active                     N   
4  ...     FH,YW,YZ   IBRD  Active                     N   

                                              theme1  \
0      {'Percent': 100, 'Name': 'Education for all'}   
1  {'Percent': 30, 'Name': 'Other economic manage...   
2    {'Percent': 46, 'Name': 'Regional integration'}   
3  {'Percent': 50, 'Name': 'Participation and civ...   
4  {'Percent': 30, 'Name': 'Export development an...   

                                      theme_namecode    themecode   totalamt  \
0      [{'code': '65', 'name': 'Education for all'}]           65  130000000   
1  [{'code': '24', 'name': 'Other economic manage...        54,24          0   
2  [{'code': '47', 'name': 'Regional integration'...  52,81,25,47    6060000   
3  [{'code': '57', 'name': 'Participation and civ...        59,57          0   
4  [{'code': '45', 'name': 'Export development an...        41,45   13100000   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  
2      6060000  http://www.worldbank.org/projects/P145310?lang=en  
3      1500000  http://www.worldbank.org/projects/P144665?lang=en  
4     13100000  http://www.worldbank.org/projects/P144933/seco...  

[5 rows x 50 columns]

In [5]:
# Check the columns present in the database

json_df.columns

Index(['_id', 'approvalfy', 'board_approval_month', 'boardapprovaldate',
       'borrower', 'closingdate', 'country_namecode', 'countrycode',
       'countryname', 'countryshortname', 'docty', 'envassesmentcategorycode',
       'grantamt', 'ibrdcommamt', 'id', 'idacommamt', 'impagency',
       'lendinginstr', 'lendinginstrtype', 'lendprojectcost',
       'majorsector_percent', 'mjsector_namecode', 'mjtheme',
       'mjtheme_namecode', 'mjthemecode', 'prodline', 'prodlinetext',
       'productlinetype', 'project_abstract', 'project_name', 'projectdocs',
       'projectfinancialtype', 'projectstatusdisplay', 'regionname', 'sector',
       'sector1', 'sector2', 'sector3', 'sector4', 'sector_namecode',
       'sectorcode', 'source', 'status', 'supplementprojectflg', 'theme1',
       'theme_namecode', 'themecode', 'totalamt', 'totalcommamt', 'url'],
      dtype='object')

### 1. Top 10 Countries with the Most Projects 

In [6]:
# Find the top 10 country with most projects

top10_country = json_df.groupby('countryshortname').size().sort_values(ascending=False).head(10)
top10_country

countryshortname
Indonesia             19
China                 19
Vietnam               17
India                 16
Yemen, Republic of    13
Nepal                 12
Bangladesh            12
Morocco               12
Mozambique            11
Africa                11
dtype: int64

In [7]:
# Since Africa is not a country and is included in the top 10, we need to clean up the data
# Checking what country should be used other than Africa by reading the project's description

africa = json_df[json_df.countryname == 'Africa']
africa.url

45     http://www.worldbank.org/projects/P125018/west...
46     http://www.worldbank.org/projects/P118213/rcip...
51     http://www.worldbank.org/projects/P130888/buil...
58     http://www.worldbank.org/projects/P144902?lang=en
65     http://www.worldbank.org/projects/P075941/nels...
99     http://www.worldbank.org/projects/P145160/addi...
167    http://www.worldbank.org/projects/P123093/west...
184    http://www.worldbank.org/projects/P120370/sout...
287    http://www.worldbank.org/projects/P094183/agri...
353    http://www.worldbank.org/projects/P130694/nile...
449    http://www.worldbank.org/projects/P130174/firs...
Name: url, dtype: object

In [8]:
#Change the 'countryshortname' to match the country where the project is located

# Row 45 - Multiple West African countries -> cannot be changed to a single country
# Row 46 - Change to Uganda
json_df.loc[46, 'countryshortname'] = 'Uganda'
# Row 51 - Multiple West African countries -> cannot be changed to a single country
# Row 58 - Multiple West African countries -> cannot be changed to a single country
# Row 65 - Multiple West African countries -> cannot be changed to a single country
# Row 99 - Change to Mali
json_df.loc[99, 'countryshortname'] = 'Mali'
# Row 167 - Change to Togo
json_df.loc[167, 'countryshortname'] = 'Togo'
# Row 184 - Change to Tanzania
json_df.loc[184, 'countryshortname'] = 'Tanzania'
# Row 287 - Multiple West African countries -> cannot be changed to a single country
# Row 353 - Multiple West African countries -> cannot be changed to a single country
# Row 449 - Change to Niger
json_df.loc[449, 'countryshortname'] = 'Niger'

In [9]:
# Find the top 10 countries with the most projects

df_removeAfrica = json_df[json_df.countryname != 'Africa']
top10_country = df_removeAfrica.groupby('countryshortname').size().sort_values(ascending=False).head(10)
top10_country

countryshortname
Indonesia             19
China                 19
Vietnam               17
India                 16
Yemen, Republic of    13
Morocco               12
Bangladesh            12
Nepal                 12
Mozambique            11
Brazil                 9
dtype: int64

After cleaning the data, above are the results of the top 10 countries with the most projects

### 2. Top 10 Major Project Themes

In [10]:
# Check the column 'mjtheme_namecode' to find the Major Project Themes

json_df.mjtheme_namecode.head(10)

0    [{'code': '8', 'name': 'Human development'}, {...
1    [{'code': '1', 'name': 'Economic management'},...
2    [{'code': '5', 'name': 'Trade and integration'...
3    [{'code': '7', 'name': 'Social dev/gender/incl...
4    [{'code': '5', 'name': 'Trade and integration'...
5    [{'code': '6', 'name': 'Social protection and ...
6    [{'code': '2', 'name': 'Public sector governan...
7    [{'code': '11', 'name': 'Environment and natur...
8    [{'code': '10', 'name': 'Rural development'}, ...
9    [{'code': '2', 'name': 'Public sector governan...
Name: mjtheme_namecode, dtype: object

In [11]:
# Create tables for the nested elements of 'mjtheme_namecode'

namecode = json_normalize(data, 'mjtheme_namecode')
namecode.head()

code                                   name
0    8                      Human development
1   11                                       
2    1                    Economic management
3    6  Social protection and risk management
4    5                  Trade and integration

In [12]:
# Find the top 10 Project Themes

namecode.groupby('name').size().sort_values(ascending=False)

name
Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
                                                122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
Economic management                              33
Rule of law                                      12
dtype: int64

According to the result above, we saw that there are 122 blank project's names but have project's codes. In the next section, we will replace the blanks with the project's codes.

### 3. Fill the Missing Name based on code 

In [13]:
# Find how many types of project's codes

lencode = namecode.code.unique()
n = len(lencode)
n

11

In [14]:
# Create a lookup tables by project's codes

lookup = []
for num in range(1,n+1):
    proj_name = namecode.name[namecode.code == str(num)].unique()
    if proj_name[0] == '':
        #row.append(num)
        lookup.append([num, proj_name[1]])
    else:
        #row.append(num)
        lookup.append([num, proj_name[0]])

cols = ['code', 'name']
df_lookup = pd.DataFrame(lookup, columns = cols)
df_lookup.code = df_lookup.code.astype(np.int64)
df_lookup

code                                          name
0      1                           Economic management
1      2                      Public sector governance
2      3                                   Rule of law
3      4      Financial and private sector development
4      5                         Trade and integration
5      6         Social protection and risk management
6      7                   Social dev/gender/inclusion
7      8                             Human development
8      9                             Urban development
9     10                             Rural development
10    11  Environment and natural resources management

In [15]:
# Change the project's codes to integer

namecode.head()
namecode.code = namecode.code.astype(np.int64)

In [16]:
# Merge two dataframes by keeping the code from the namecode DataFrame

results = namecode.merge(df_lookup, on='code', how='left')
results.columns

Index(['code', 'name_x', 'name_y'], dtype='object')

In [17]:
# Remove the columns with blank project's names

results = results[['code', 'name_y']].copy()
results = results.rename(columns = {'name_y' : 'name'})
results.head()

code                                          name
0     8                             Human development
1    11  Environment and natural resources management
2     1                           Economic management
3     6         Social protection and risk management
4     5                         Trade and integration

In [18]:
# Group by the data by project's names and find the top 10

results.groupby('name').size().sort_values(ascending=False).head(10)

name
Environment and natural resources management    250
Rural development                               216
Human development                               210
Public sector governance                        199
Social protection and risk management           168
Financial and private sector development        146
Social dev/gender/inclusion                     130
Trade and integration                            77
Urban development                                50
Economic management                              38
dtype: int64

### Not Used 

In [19]:
## Unused code (for future references)

# lookup = {}
# for num in range(1,n+1):
#     proj_name = namecode.name[namecode.code == str(num)].unique()
#     if proj_name[0] == '':
#         case = {num : proj_name[1]}
#         lookup.update(case)
#     else:
#         case = {num : proj_name[0]}
#         lookup.update(case)
    
# lookup